cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -427,20 +434,39 @@ public boolean apply(ProgramRuntimeService.RuntimeInfo runtimeInfo) {
    * @throws Exception
    */
   public void removeApplication(final ApplicationId appId) throws Exception {
+    findRunningProgram(appId);
+    ApplicationSpecification spec = store.getApplication(appId);
+    if (spec == null) {
+      throw new NotFoundException(appId);
+    }
+    // if the application has only one version, do full deletion, else only delete the specified version
+    if (store.getAllAppVersions(appId).size() == 1) {
+      deleteApp(appId, spec);
+      return;
+    }
+    deleteAppVersion(appId, spec);
+  }
+
+  /**
+   * Find if the given application has running programs
+   *
+   * @param appId the id of the application to find running programs for
+   * @throws CannotBeDeletedException : the application cannot be deleted because of running programs
+   */
+  private void findRunningProgram(final ApplicationId appId) throws CannotBeDeletedException {
```

Comment:
you're not just finding right, you're bailing if there are running programs. Call it ``ensureNoRunningPrograms`` or similar